### PR TITLE
JS: Update the extractor to use TypeScript 4.7

### PR DESCRIPTION
**Pull request by kaeluka**
_Erik here:_

A plain update to TypeScript 4.7 without surprises.  

There is not much special happening, but TypeScript now supports `.mts` and `.cts` files, so we had to add those to the extractor.  

Tests have been added that we can handle all the features mentioned in [the change-notes](https://devblogs.microsoft.com/typescript/announcing-typescript-4-7/).  

I can't test it locally, but the tests passed on the release candidate, so I highly expect them to pass on this stable release.   
The tests will fail here until the internal PR has been merged.

### CODE REVIEW

File: javascript/extractor/lib/typescript/package.json
Comment (kaeluka) [inline javascript/extractor/lib/typescript/package.json:5]: this will be updated to 4.7.2 or something like that

File: javascript/extractor/lib/typescript/yarn.lock
Comment (kaeluka) [inline javascript/extractor/lib/typescript/yarn.lock:10]: will be updated as well

**Review by erik-krogh — COMMENTED** [review]
I don't see a test/implementation of `moduleSuffixes`.  
It can wait until after this PR has been merged, but we should consider it.  
https://devblogs.microsoft.com/typescript/announcing-typescript-4-7-rc/#resolution-customization-with-modulesuffixes 

It probably requires some changes in `NodeModuleResolutionImpl.qll` around the `tryExtensions` predicate.  

---- 

Otherwise LGTM, with two small comments from a quick look.

**Comment by kaeluka** [comment]
I've now added a test for the moduleSuffixes feature as well!

**Comment by erik-krogh** [comment]
> I've now added a test for the moduleSuffixes feature as well!

You've tested that TypeScript can figure out the types, but not that our own module resolution can figure out the imports.  
(See `NodeModuleResolutionImpl.qll`).  
I can wait until after this PR has been merged.

**Comment by kaeluka** [comment]
Yes, that's what I did :-) i know there's still work left after merging :+1:

**Comment by erik-krogh** [comment]
I've bumped the version to the stable release that just happened.  
It would be nice to get someone else to review it.

File: javascript/ql/test/library-tests/TypeScript/Types/tst.ts
Comment (erik-krogh) [reply javascript/ql/test/library-tests/TypeScript/Types/tst.ts:325]: The `Map<string,Error>` is an `ExpressionWithTypeArguments` from `TypeScript.qll`.  

But we did not show that in the printAst output, so I fixed that.

File: javascript/ql/test/library-tests/TypeScript/Types/tst.ts
Comment (asgerf) [inline javascript/ql/test/library-tests/TypeScript/Types/tst.ts:332]: Do we extract the `extends string` part?

Usually we've tried to extract enough of the AST to prevent spurious unused variable/import alerts. Since the `extends string` part could instead contain a reference to an imported name, we'd want to extract it.

The unused variable alert isn't that important anymore, but I still think we should try to maintain it.

Comment (erik-krogh) [reply javascript/ql/test/library-tests/TypeScript/Types/tst.ts:332]: Yes we extract it, and I've added a test that it works with our unused variable query.

**Comment by erik-krogh** [comment]
> Thanks for addressing my questions @erik-krogh!
> 
> Everything LGTM, but could you please run an evaluation before merging?

[Evaluation was uneventful](https://github.com/github/codeql-dca-main/tree/data/erik-krogh/pr-9235-62fd3fd__nightly-old__code-scanning/reports).  
We extract one more LOC in Babel. I suppose that comes from a TS test-case.

**Review by asgerf — APPROVED** [review]
Great! LGTM

**Pull request merged**
